exercises/sat/descr.md

```diff
@@ -0,0 +1,616 @@
+have been recorded so far. Note that a clause that contains both the literal
+`x` and the opposite literal `~x` is unsatisfiable. In such a situation, the
+function `new_clause` must raise the exception `UNSAT`. *Hint:* to detect this
+situation, use `intersect`.
+
+**Question C.** (Ungraded.) Implement the function `new_var`, which creates and
+returns a new variable.
+
+## The Undo Trail
+
+A SAT solver searches a tree of candidate solutions to a Boolean constraint.
+Repeatedly, it picks a variable `x` whose value has not yet been decided, sets
+this variable to `true` (for instance), and explores the consequences of this
+decision, which form a subtree. If it appears after a while that this decision
+was wrong (that is, there is no solution in this subtree) then the algorithm
+must *backtrack*, that is, come back to the point where this decision was
+made, undo this decision, make the opposite decision (that is, set `x` to
+`false`), and explore the consequences of this new decision, which form
+another subtree.
+
+A solver usually maintains several mutable data structures, such as a set of
+undecided variables, a set of unit clauses, etc. When the solver backtracks to
+an earlier point in time, it must reestablish an earlier state of these data
+structures. This is typically implemented either by saving and restoring an
+entire state or by keeping track of an *undo trail*, that is, a stack of *undo
+actions* that must be performed in order to return to earlier states.
+
+An undo trail presents the following abstract interface:
+
+```
+  module Trail () : sig
+    val push: (unit -> unit) -> unit
+    type checkpoint
+    val record: unit -> checkpoint
+    val revert: checkpoint -> unit
+  end
+```
+
+The module `Trail` is in fact a functor: that is, it takes an empty argument
+`()`. The functor application `Trail()` initializes a new undo trail, which
+is initially empty. This undo trail supports three operations:
+
+* `push action` pushes a new undo action, represented by a function of type
+  `unit -> unit`, onto the trail.
+
+  The undo trail has no knowledge of what this undo action does, or what data
+  structures it affects. For instance, perhaps the user has just incremented a
+  counter `cnt` of type `int ref` and wishes to push an undo action whose
+  effect is to decrement `cnt`. The undo trail doesn't need to know; it takes
+  care only of executing the right undo actions at the right time.
+
+* `record()` returns the current point in time.
+  Such a value is known as a *checkpoint*.
+
+* `revert c` goes back in time to the checkpoint `c`
+  by executing *in reverse order* all of the undo actions
+  that were pushed after the checkpoint `c` was recorded.
+
+  As a side effect, this invalidates all checkpoints that
+  are strictly newer than `c`; these checkpoints must no
+  longer be used.
+
+**Question 6.** Implement the module `Trail`.
+
+*Hint.* Maintain a stack of undo actions, stored in an infinite array.
```

Comment:
I guess that you are advising for an infinite array instead of a reference to a list for a good reason. Is it only for efficiency?

Reply:
Efficiency was not a primary concern. Simplicity was. I thought using an array was the simplest solution, but you are right, there is probably an equally simple solution based on a reference to a list.

Reply:
OK. It is true that an infinite array is arguably easy to use.